### PR TITLE
Launch GameScope/Steam session from Display Manager

### PR DESCRIPTION
**Pull request by nrdxp**
###### Description of changes

With the new option, `programs.steam.gamescopeSession` one can add a session file to load a Steam Big Picture session directly from a display manager using GameScope as the window manager.

> Note:
> A SteamOS session was deliberately not used since there is no way to exit the session from the UI, making for bad UX if user wants to switch to another session.

There is also a new `programs.gamescope` module to configure gamescope appropriately. For example, I have an Optimus laptop where the Nvidia GPU is wired directly to the HDMI port, meaning I cannot launch gamescope directly with this GPU unless I am using an external monitor. If I wanna play on my laptop screen using the stronger GPU I have to configure gamescope like so:
```nix
{
  programs.gamescope = {
    enable = true;    
    args = ["--rt" "--prefer-vk-device 8086:9bc4"];
    env = {
      __NV_PRIME_RENDER_OFFLOAD = "1";
      __VK_LAYER_NV_optimus = "NVIDIA_only";
      __GLX_VENDOR_LIBRARY_NAME = "nvidia";
    };
  };
}
```
where `8086:9bc4` is the hardware id for the Intel GPU, so that gamescope itself renders to the laptop screen, but the offload variables are set so that any programs launched (i.e. Steam) will use the Nvidia one.

There is also a `programs.gamescope.capSysNice` option to allow gamescope to renice itself. To use this with Steam the bubblewrap binary has to have the setuid bit, so included is an override when this is enabled to do just that in the Steam derivation.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/steam.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:30]: Please use makeDesktopItem

Comment (nrdxp) [reply nixos/modules/programs/steam.nix:30]: I'm not sure if that's correct in this case, since the desktop file must live at "share/wayland-sessions/steam.desktop", and `makeDesktopItem` hardcodes ["share/applications"](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/make-desktopitem/default.nix#L115) which is useless for `services.xserver.displayManager.sessionPackages`.

Although, perhaps it would be useful to modify `makeDesktopItem` to allow for overriding the destination?

Comment (SuperSandro2000) [reply nixos/modules/programs/steam.nix:30]: > services.xserver.displayManager.sessionPackages

Do you know of any other usecases that need a different place for the desktop file? If there are multiple we should definitely add an option to change the directory otherwise I am a bit undecided.

Comment (nrdxp) [reply nixos/modules/programs/steam.nix:30]: Just this one option that I'm aware of but, there are two possible locations "share/wayland-sessions" and "share/xsessions"

**Comment by rapenne-s** [comment]
It may be worth adding a `programs.gamescope.optimus = true` so people don't have to figure the environment variables puzzle?

**Comment by rapenne-s** [comment]
why is it `programs.gamescope` and not `services.xserver.windowManager.gamescope`? The latter sounds fitting more to me, given it's a new choice for a session at login time.

**Comment by rapenne-s** [comment]
I'm using sddm, when I log in with my user and the Steam (wayland) session, the display-manager service is restarting instantly.

Do you know how I could find useful debug information? `journalctl -u display-manager.service` contains a lot of lines and I can't find anything significant wrt gamescope (I suppose it crashes) at login

**Comment by nrdxp** [comment]
> why is it `programs.gamescope` and not `services.xserver.windowManager.gamescope`? The latter sounds fitting more to me, given it's a new choice for a session at login time.

It doesn't launch using X as a window manager, instead it is a wl-roots based wayland compositor. For example, sway is also a similarly based project and it is also configured through `programs.sway`.

File: nixos/modules/programs/gamescope.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/gamescope.nix:8]: Won't this require us to recompile the package? That doesn't seem like a great user experience. 

I think the best solution is double wrapping. That's also not great, but its better than having to recompile the program whenever you change one of the flags.

File: nixos/modules/programs/gamescope.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/gamescope.nix:15]: Can you add a `package` option? That should allow for overriding the `gamescope` derivation this module uses more easily.

File: nixos/modules/programs/gamescope.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/gamescope.nix:59]: Does the security wrapper add it to `PATH` if `sysCapNice` is enabled?

File: nixos/modules/programs/steam.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/steam.nix:79]: Might be useful to add a comment explaining what this flag does.

File: nixos/modules/programs/steam.nix
Comment (IvarWithoutBones) [reply nixos/modules/programs/steam.nix:30]: Having the option to override the location of the produced desktop file definitely seems useful to me, even if nothing else in nixpkgs uses it now.

File: nixos/modules/programs/gamescope.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/gamescope.nix:38]: This example does seem pretty useful for people with nvidia laptops, maybe we could have an `exampleText` with a comment explaining what it does?

File: nixos/modules/programs/gamescope.nix
Comment (nrdxp) [reply nixos/modules/programs/gamescope.nix:8]: I considered this, originally I decided it was alright because the package is actually surprisingly light to build, but I can see your point. Perhaps we should double wrap :thinking:

File: nixos/modules/programs/gamescope.nix
Comment (nrdxp) [reply nixos/modules/programs/gamescope.nix:59]: yes it does

File: nixos/modules/programs/steam.nix
Comment (zhaofengli) [reply nixos/modules/programs/steam.nix:79]: We should probably use the long-form flag `--steam` along with a comment.

File: nixos/modules/programs/gamescope.nix
Comment (IvarWithoutBones) [reply nixos/modules/programs/gamescope.nix:8]: I can understand that, though for some shell arguments I'd argue any time spend compiling is too much.

**Comment by IvarWithoutBones** [comment]
@nrdxp are you still interested in this? I'd love to see it get merged. If you're not I can pick up where you left off, while retaining your authorship on the commits of course.

**Comment by nrdxp** [comment]
> @nrdxp are you still interested in this? I'd love to see it get merged. If you're not I can pick up where you left off, while retaining your authorship on the commits of course.

Thank you. That is fine by me, I may have some time this weekend to address some things here, but if not feel free to hop in.

**Comment by nrdxp** [comment]
@IvarWithoutBones, I had some time to finally address some of the things in the review. Hopefully we can merge soon.

**Comment by nrdxp** [comment]
> It may be worth adding a `programs.gamescope.optimus = true` so people don't have to figure the environment variables puzzle?

This isn't a bad suggestion, but it also seems a little to specific, since optimus is not the only possible prime setup. There are also different ways to run gamescope depending on preference. For example, passing the `prefer-vk-device` to your intel chip while also setting the prime offload vars will allow gamescope to start on your laptop screen but running with the stronger dgpu. Something this might be what's desired, other times you might want it to launch on an external screen.

Not sure if it is worth all the trouble or not to capture all the possible configurations.

File: nixos/modules/programs/steam.nix
Comment (IvarWithoutBones) [inline nixos/modules/programs/steam.nix:7]: I think this should be a bit more descriptive since this derivation mostly concerns just steam itself
```suggestion
  gamescopeCfg = config.programs.gamescope;
```

**Comment by IvarWithoutBones** [comment]
> @IvarWithoutBones, I had some time to finally address some of the things in the review. Hopefully we can merge soon. Also, I added the same args and env options available on programs.gamescope to steam.gamescopeSession so that one can set vars and args for the session that they don't necessarily want for gamescope launched elsewhere.

Sounds very useful, thanks! I just have one very minor comment, other than that this looks good 👍 

> Not sure if it is worth all the trouble or not to capture all the possible configurations.

Agreed, documenting how the user can do this themselves (as done in this PR) is more maintainable than covering all edge cases ourselves.

File: nixos/modules/programs/steam.nix
Comment (nrdxp) [reply nixos/modules/programs/steam.nix]: done

**Review by rapenne-s — APPROVED** [review]
works fine for me <3

**Comment by rapenne-s** [comment]
oh, it seems there is an issue when you don't use gamescope

```
error: The option `programs.steam.gamescopeSession' is used but not defined.
(use '--show-trace' to show detailed location information)
```

**Comment by nrdxp** [comment]
I'll see if I can track that down.

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:101]:
```suggestion
    programs.gamescope.enable = mkDefault cfg.gamescopeSession.enable;
```

Comment (nrdxp) [reply nixos/modules/programs/steam.nix]: I opted for:
```nix
{
    programs.gamescope.enable = mkIf cfg.gamescopeSession.enable (mkDefault true);
}
```

Comment (Atemu) [reply nixos/modules/programs/steam.nix]: Nit: I find it a bit weird to have true behind a condition. Might as well just have the condition in there directly.

Comment (nrdxp) [reply nixos/modules/programs/steam.nix]: yeah you are right, I must not have been thinking correctly :sweat_smile:

**Comment by PedroHLC** [comment]
What is the current status of this PR?

**Comment by nrdxp** [comment]
I've been using it on my system since I authored it and it seems to have worked fine so far. Haven't had time to use it for a while though. Looks like I gotta merge conflict to work out. I'll see if I can adress that soon, and then we can probably merge. I just never got an approval so I help off, but I've been using it for quite a while now so I think it should be fine.

File: nixos/modules/programs/steam.nix
Comment (PedroHLC) [inline nixos/modules/programs/steam.nix:11]: this is the line that was updated and needs a rebase.

File: nixos/modules/programs/steam.nix
Comment (PedroHLC) [inline nixos/modules/programs/steam.nix:28]: Can we expose this binary to systemPackages to have it in PATH for users without DMs?

**Review by PedroHLC — COMMENTED** [review]
Like mentioned in https://github.com/NixOS/nixpkgs/pull/187507#issuecomment-1353161605, having `programs.gamescope.enable` and not `programs.steam.gamescopeSession.enable` results in error.

File: nixos/modules/programs/steam.nix
Comment (PedroHLC) [reply nixos/modules/programs/steam.nix:28]: Ok, I did it here: https://github.com/PedroHLC/nixpkgs/commit/4e4c9644270ef21770c8d835a36da1c81b1c7bb1.

Opened TTY3, logged in, typed `steam-gamescope`, and this thing is working amazingly.

**Comment by thiagokokada** [comment]
> I've been using it on my system since I authored it and it seems to have worked fine so far. Haven't had time to use it for a while though. Looks like I gotta merge conflict to work out. I'll see if I can adress that soon, and then we can probably merge. I just never got an approval so I held off, but I've been using it for quite a while now so I think it should be fine.

The PR looks fine to me, if you can fix the merge conflict I can go approve/merge it.

File: nixos/modules/programs/gamescope.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/gamescope.nix:39]: This does not explain when I I would want to set this option.

File: nixos/modules/programs/steam.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:17]:
```suggestion
  steam-gamescope = let
    exports = builtins.attrValues (builtins.mapAttrs (n: v: "export ${n}=${v}") cfg.gamescopeSession.env);
  in
    pkgs.writeShellScriptBin "steam-gamescope" ''
      ${builtins.concatStringsSep "\n" exports}
      gamescope --steam ${toString cfg.gamescopeSession.args} -- steam -tenfoot -pipewire-dmabuf
    '';
```

File: nixos/modules/programs/steam.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:72]: Can you please revert the formatting changes here? They make reviewing harder and I don't actually know if something here changed or not on a glance.

File: nixos/modules/programs/steam.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:132]: What is the error? Can we maybe reduce this to a few caps instead?

File: nixos/modules/programs/steam.nix
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:162]: Please revert the formatting changes.

File: nixos/modules/programs/steam.nix
Comment (nrdxp) [reply nixos/modules/programs/steam.nix:118]: left a more discriptive comment

**Comment by nrdxp** [comment]
@SuperSandro2000, apologies, I didn't realize I applied a formatter to the file, I fixed them all and address your other comments. This should be ready to merge. If there are no objections I'll merge in a few days.

File: nixos/modules/programs/steam.nix
Comment (nrdxp) [reply nixos/modules/programs/steam.nix]: I added this in the latest rebase.

**Comment by oxalica** [comment]
This breaks my config on master. I only has `programs.steam.enable = true;` and it fails with ``error: The option `programs.steam.gamescopeSession' is used but not defined.``. This is exactly the same as https://github.com/NixOS/nixpkgs/pull/187507#issuecomment-1353161605 which is ignored somehow.

**Comment by nrdxp** [comment]
Hmm actually I thought I fixed that, my mistake, I'll prepare a fix.

**Comment by oxalica** [comment]
> Hmm actually I thought I fixed that, my mistake, I'll prepare a fix.

IMO, we should at least have some tests and approvals from others before self-merging.

**Comment by nrdxp** [comment]
Here is the fix:
https://github.com/NixOS/nixpkgs/pull/228571

I have been using this module since gamescope first was released to the public, so I have tested it extensively, but not with the session disabled which is how I overlooked it. In any case I've waited for an approval for over 6 months and we've had several review rounds already. In any case, this should resolve things.

**Pull request merged**
